src/modules/netinstall/PackageModel.cpp

```diff
@@ -0,0 +1,260 @@
+ */
+
+#include "PackageModel.h"
+
+#include "utils/Logger.h"
+#include "utils/YamlUtils.h"
+
+#include <Qt>
+
+PackageModel::PackageModel( const YAML::Node& data, const QVariantList& columnHeadings,
+                            QObject* parent ) :
+    QAbstractItemModel( parent ),
+    m_columnHeadings( columnHeadings )
+{
+    QVariantList rootData;
+    m_rootItem = new PackageTreeItem( );
+    setupModelData( data, m_rootItem );
+}
+
+PackageModel::~PackageModel()
+{
+    delete m_rootItem;
+}
+
+QModelIndex
+PackageModel::index( int row, int column, const QModelIndex& parent ) const
+{
+    if ( !hasIndex( row, column, parent ) )
+        return QModelIndex();
+
+    PackageTreeItem* parentItem;
+
+    if ( !parent.isValid() )
+        parentItem = m_rootItem;
+    else
+        parentItem = static_cast<PackageTreeItem*>( parent.internalPointer() );
+
+    PackageTreeItem* childItem = parentItem->child( row );
+    if ( childItem )
+        return createIndex( row, column, childItem );
+    else
+        return QModelIndex();
+}
+
+QModelIndex
+PackageModel::parent( const QModelIndex& index ) const
+{
+    if ( !index.isValid() )
+        return QModelIndex();
+
+    PackageTreeItem* child = static_cast<PackageTreeItem*>( index.internalPointer() );
+    PackageTreeItem* parent = child->parentItem();
+
+    if ( parent == m_rootItem )
+        return QModelIndex();
+    return createIndex( parent->row(), 0, parent );
+}
+
+int
+PackageModel::rowCount( const QModelIndex& parent ) const
+{
+    if ( parent.column() > 0 )
+        return 0;
+
+    PackageTreeItem* parentItem;
+    if ( !parent.isValid() )
+        parentItem = m_rootItem;
+    else
+        parentItem = static_cast<PackageTreeItem*>( parent.internalPointer() );
+
+    return parentItem->childCount();
+}
+
+int
+PackageModel::columnCount( const QModelIndex& parent ) const
+{
+    if ( parent.isValid() )
+        return static_cast<PackageTreeItem*>( parent.internalPointer() )->columnCount();
+    return m_rootItem->columnCount();
+}
+
+QVariant
+PackageModel::data( const QModelIndex& index, int role ) const
+{
+    if ( !index.isValid() )
+        return QVariant();
+    PackageTreeItem* item = static_cast<PackageTreeItem*>( index.internalPointer() );
+    if ( index.column() == 0 && role == Qt::CheckStateRole )
+        return item->isSelected();
+
+    if ( !item->childCount() ) // package
+    {
+        if ( !item->parentItem()->isHidden() && role == Qt::DisplayRole &&
+                index.column() == 0 )
+            return QVariant( item->packageName() );
+        else if ( role == PackageTreeItem::PackageNameRole )
+            return QVariant( item->packageName() );
+        else
+            return QVariant();
+    }
+
+    if ( item->isHidden() && role == Qt::DisplayRole ) // Hidden group
+        return QVariant();
+
+    switch ( role )
+    {
+    case Qt::DisplayRole:
+        return item->data( index.column() );
```

Comment:
Because the item holds the data not the model? But I do see your point with the rest of the function's behaviour